poster/src/poster.ts

```diff
@@ -31,195 +42,121 @@ async function main(sources: string,
+    payloads: OpenPriceFeedPayload[],
+    viewAddress: string,
+    supportedAssets: string[],
+    delta: number,
+    web3: Web3): Promise<OpenPriceFeedPayload[]> {
 
-      const prev_price = await helpers.getPreviousPrice(sourceAddress, asset, dataAddress, web3);
-      console.log(`For asset ${asset}: prev price = ${prev_price}, new price = ${price}`);
+  const dataAddress = await getDataAddress(viewAddress, web3);
 
-      // Update price only if new price is different by more than delta % from the previous price
-      if (!inDeltaRange(delta, Number(price), prev_price)) {
-        console.log(`Not in delta range for asset ${asset}, price = ${Number(price)}, prev price = ${prev_price}`);
-        filteredPrices[asset] = price;
-        filteredMessages.push(payload.messages[index]);
-        filteredSignatures.push(payload.signatures[index]);
-      }
-      index++;
+  let filteredPayloads = await Promise.all(payloads.map(async payload => {
+    let sourceAddresses = await Promise.all(payload.messages.map((_, i) => {
+      return getSourceAddress(dataAddress, payload.messages[0], payload.signatures[0], web3);
+    }));
+
+    if ([...new Set(sourceAddresses)].length !== 1) {
+      throw new Error(`Invalid source addresses, got: ${JSON.stringify(sourceAddresses)}`);
     }
-    payload.prices = filteredPrices;
-    payload.messages = filteredMessages;
-    payload.signatures = filteredSignatures;
-  }))
-
-  // Filter payloads with no updated prices
-  const filteredPayloads = payloads.filter(payload => payload.messages.length > 0);
-  return filteredPayloads;
+    const sourceAddress = sourceAddresses[0]; // We proved they all match this single address
+
+    return await Object.entries(payload.prices).reduce<Promise<OpenPriceFeedPayload>>(async (accP, [asset, priceRaw]: [string, string], index: number) => {
+      let acc = await accP;
+      let price = Number(priceRaw);
+
+      if (supportedAssets.includes(asset.toUpperCase())) {
+        const prevPrice = await getPreviousPrice(sourceAddress, asset, dataAddress, web3);
+
+        console.debug(`Previous Price Data: asset=${asset}, prev_price=${prevPrice}, new_price=${price}`);
+
+        // Update price only if new price is different by more than delta % from the previous price
+        if (!inDeltaRange(delta, price, prevPrice)) {
+          console.log(`Setting Price: asset=${asset}, price=${price}, prev_price=${prevPrice}`);
+
+          return {
+            prices: {
+              ...acc.prices,
+              [asset]: priceRaw
+            },
+            messages: [...acc.messages, payload.messages[index]],
+            signatures: [...acc.signatures, payload.signatures[index]]
+          };
+        } else {
+          return acc;
+        }
+      } else {
+        // Post only prices for supported assets, skip prices for unregistered assets
+        console.debug(`Skipping ${asset} as not part of supported assets: ${JSON.stringify(supportedAssets)}`);
```

Comment:
Nit, may be more clear to filter the assets before or after this reduce action like
```
Object.entries(payload.prices).filter( asset => supportedAssets.includes(asset.toUpperCase())).reduce...
```
loses the logging though 🤷